### PR TITLE
Add cli.Command that parses os.Args values into the config

### PR DESCRIPTION
**Pull request by AlexisMontagne**
### What does this PR do?

It allows the user to use cli args as yet another source of config population

### What are the observable changes?

```go
package main

import (
	"context"
	"fmt"

	"github.com/upfluence/cfg/x/cli"
)

type config struct {
	Foo string `flag:"foo,f"`
	Bar string `flag:"bar,b"`
	Buz string `arg:"buz"`
}

func main() {
	cli.NewApp(
		cli.WithName("example"),
		cli.WithCommand(
			cli.ArgumentCommand{
				Variable: "buz",
				Command: cli.StaticCommand{
					Help:     cli.HelpWriter(&config{}),
					Synopsis: cli.SynopsisWriter(&config{}),
					Execute: func(ctx context.Context, cctx cli.CommandContext) error {
						var c config

						if err := cctx.Configurator.Populate(ctx, &c); err != nil {
							return err
						}

						fmt.Printf("config: %+v", c)

						return nil
					},
				},
			},
		),
	).Run(context.Background())
}

// exampe foo -f buz
// config {Foo:buz Bar: Buz:foo}

// exampe -f buz
// no argument found for variable buz
```

### Good PR checklist

- [x] Title makes sense
- [x] Is against the correct branch
- [x] Only addresses one issue
- [x] Properly assigned
- [x] Added/updated tests
- [ ] Added/updated documentation
- [x] Properly labeled

### Additional Notes

As it stands the ArgumentCommand does not impact neither the Synopsis nor the Help message. I think we should pass along a stack of (parsed args and parsed config) across the command wrapping. However to do that we need to break some public type definition.

It will be done in a future PR

### CODE REVIEW

File: x/cli/argument_command.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+func (sc ArgumentCommand) WriteHelp(w io.Writer) (int, error) {
+	return sc.Command.WriteHelp(w)
+}
+
+func (sc ArgumentCommand) Run(ctx context.Context, cctx CommandContext) error {
+	if len(cctx.Args) == 0 {
+		if _, err := fmt.Fprintf(
+			cctx.Stderr,
+			"no argument found for variable %q, follow the synopsys:\n",
```
Comment (jlevesy) [inline x/cli/argument_command.go]:
```suggestion
			"no argument found for variable %q, follow the synopsis:\n",
```

File: x/cli/argument_command.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package cli
+
+import (
+	"context"
+	"fmt"
+	"io"
+)
+
+type ArgumentCommand struct {
```
Comment (jlevesy) [inline x/cli/argument_command.go:9]:
```suggestion
type Argument struct {
```

or

```suggestion
type CommandArgument struct {
```

WDYT ?

File: x/cli/app.go [hunk 3]
```diff
@@ -79,16 +79,22 @@ func (a *App) parseArgs() ([]string, []string) {
 }
 
 func (a *App) commandContext() CommandContext {
-	cmds, flags := a.parseArgs()
+	var (
+		cmds, flags = a.parseArgs()
+		args        = make(map[string]string)
```
Comment (jlevesy) [inline x/cli/app.go:84]:
```suggestion
		args        = make(argProvider)
```
?

File: x/cli/help_command.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+const defaultHelp = "no help content provided"
+
+type helpCommand struct {
+	cmd Command
+}
+
+func (hc *helpCommand) WriteHelp(w io.Writer) (int, error) {
+	return io.WriteString(w, "Print this message")
```
Comment (jlevesy) [inline x/cli/help_command.go:15]:
```suggestion
	return io.WriteString(w, "Prints this message")
```

File: x/cli/sub_command.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	"fmt"
+	"io"
+	"sort"
+	"text/tabwriter"
+)
+
+type SubCommand map[string]Command
+
+func (sc SubCommand) WriteHelp(w io.Writer) (int, error) {
+	n, err := io.WriteString(w, "Sub commands available: \n")
```
Comment (jlevesy) [inline x/cli/sub_command.go]:
```suggestion
	n, err := io.WriteString(w, "Available sub commands: \n")
```

File: x/cli/version_command.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+type versionCommand struct {
+	name    string
+	version string
+}
+
+func (vc *versionCommand) WriteSynopsis(io.Writer) (int, error) { return 0, nil }
+
+func (vc *versionCommand) WriteHelp(w io.Writer) (int, error) {
+	return io.WriteString(w, "Print the app version")
```
Comment (jlevesy) [inline x/cli/version_command.go:19]:
```suggestion
	return io.WriteString(w, "Prints the app version")
```

**Review by jlevesy — COMMENTED** [review]
:eyes:

File: x/cli/version_command.go (hunk 6 above)
Comment (Sypheos) [reply x/cli/version_command.go:19]: Do we have to put an `s` ? It's the imperative form, isn't it ?

Comment (jlevesy) [reply x/cli/version_command.go:19]: I would say third person, "this command prints the app version",  but that might be my broken angliche talking :)

**Review by Sypheos — COMMENTED** [review]
lgtm, just some of the comments above

File: x/cli/version_command.go (hunk 6 above)
Comment (Sypheos) [reply x/cli/version_command.go:19]: I think the imperative form would fit better as those are commands you want to execute (like an order). It is also the standard form in all the helpers I know.

Comment (jlevesy) [reply x/cli/version_command.go:19]: Okay nvn

File: x/cli/app.go (hunk 3 above)
Comment (AlexisMontagne) [reply x/cli/app.go:84]: For the commandContext args is only a `map[string]string` that represents the data parsed, argProvider is here just to be a polyfill of provider. This type does not and should not be used everywhere

File: x/cli/argument_command.go (hunk 2 above)
Comment (AlexisMontagne) [reply x/cli/argument_command.go:9]: `XyzCommand` is a pattern here to tell you what is a command, checkout the godoc it makes it very clear https://godoc.org/github.com/upfluence/cfg/x/cli (SubCommand, StaticCommand) same goes for the private ones

**Review by jlevesy — APPROVED** [review]

**Pull request merged**